README.md
# egg-typeorm

[TypeORM](https://typeorm.io/#/) plugin for Egg.js.

[![NPM version][npm-image]][npm-url]
[![npm download][download-image]][download-url]

[npm-image]: https://img.shields.io/npm/v/@forsigner/egg-typeorm.svg?style=flat-square
[npm-url]: https://npmjs.org/package/@forsigner/egg-typeorm
[download-image]: https://img.shields.io/npm/dm/@forsigner/egg-typeorm.svg?style=flat-square
[download-url]: https://npmjs.org/package/@forsigner/egg-typeorm

<!--
Description here.
-->

## Install

```bash
$ yarn add @forsigner/egg-typeorm mysql
```

## Usage

### Plugin

```ts
// {app_root}/config/plugin.ts
const plugin: EggPlugin = {
  typeorm: {
    enable: true,
    package: '@forsigner/egg-typeorm',
  },
}
```

### Configuration

```ts
// {app_root}/config/config.default.ts
config.typeorm = {
  type: 'mysql',
  host: 'localhost',
  port: 3306,
  username: 'root',
  password: '123456',
  database: 'test',
  synchronize: true,
  logging: false,
  entities: ['app/entity/**/*.ts'],
  migrations: ['app/migration/**/*.ts'],
  subscribers: ['app/subscriber/**/*.ts'],
  cli: {
    entitiesDir: 'app/entity',
    migrationsDir: 'app/migration',
    subscribersDir: 'app/subscriber',
  },
}
```

### Create entity files

```bash
├── controller
│   └── home.ts
├── entity
    ├── Post.ts
    └── User.ts
```

### Entity file

```ts
// app/entity/User.ts

import { Entity, PrimaryGeneratedColumn, Column } from 'typeorm'

@Entity()
export class User {
  @PrimaryGeneratedColumn()
  id: number

  @Column()
  name: string
}
```

### Use with FindOptions

```ts
// in controller
export default class UserController extends Controller {
  public async index() {
    const { ctx } = this
    ctx.body = await ctx.repo.User.find()
  }
}
```

### Use with QueryBuilder

```ts
// in controller
export default class UserController extends Controller {
  public async index() {
    const { ctx } = this
    const firstUser = await ctx.repo.User.createQueryBuilder('user')
      .where('user.id = :id', { id: 1 })
      .getOne()
    ctx.body = firstUser
  }
}
```

## Example

[example](https://github.com/forsigner/egg-typeorm/tree/master/example)

## Questions & Suggestions

Please open an issue [here](https://github.com/forsigner/egg-typeorm/issues).

## License

[MIT](LICENSE)


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [1.5.1](http://github.com/forsigner/forsigner/egg-typeorm/compare/v1.5.0...v1.5.1) (2019-10-25)


### Bug Fixes

* choose .ts or .js by checking TS loader. ([91f4fc3](http://github.com/forsigner/forsigner/egg-typeorm/commit/91f4fc3))
* unittest env require ts files ([f41692b](http://github.com/forsigner/forsigner/egg-typeorm/commit/f41692b))



# [1.5.0](http://github.com/forsigner/forsigner/egg-typeorm/compare/v1.4.0...v1.5.0) (2019-09-02)


### Features

* useContainer ([4451b50](http://github.com/forsigner/forsigner/egg-typeorm/commit/4451b50))



# [1.4.0](http://github.com/forsigner/forsigner/egg-typeorm/compare/v1.3.2...v1.4.0) (2019-08-29)


### Features

* support export {SomeEntity} ([53dbc3c](http://github.com/forsigner/forsigner/egg-typeorm/commit/53dbc3c))



## [1.3.2](http://github.com/forsigner/forsigner/egg-typeorm/compare/v1.3.1...v1.3.2) (2019-07-15)



## [1.3.1](http://github.com/forsigner/forsigner/egg-typeorm/compare/v1.3.0...v1.3.1) (2019-07-02)


### Bug Fixes

* handle window sep ([9adaeec](http://github.com/forsigner/forsigner/egg-typeorm/commit/9adaeec))



# [1.3.0](http://github.com/forsigner/forsigner/egg-typeorm/compare/v1.2.1...v1.3.0) (2019-03-25)


### Features

* model->repo ([30725ff](http://github.com/forsigner/forsigner/egg-typeorm/commit/30725ff))



## [1.2.1](http://github.com/forsigner/forsigner/egg-typeorm/compare/v1.2.0...v1.2.1) (2019-03-08)


### Bug Fixes

* require js file if prod ([a0087bb](http://github.com/forsigner/forsigner/egg-typeorm/commit/a0087bb))



# [1.2.0](http://github.com/forsigner/forsigner/egg-typeorm/compare/v1.1.0...v1.2.0) (2019-03-07)


### Features

* check typeorm config ([45b091c](http://github.com/forsigner/forsigner/egg-typeorm/commit/45b091c))
* support ctx.connection ([fce9f17](http://github.com/forsigner/forsigner/egg-typeorm/commit/fce9f17))
* support ctx.entity ([b10cef0](http://github.com/forsigner/forsigner/egg-typeorm/commit/b10cef0))



# [1.1.0](http://github.com/forsigner/forsigner/egg-typeorm/compare/v1.0.1...v1.1.0) (2019-03-06)


### Features

* check entity dir exist ([ef657da](http://github.com/forsigner/forsigner/egg-typeorm/commit/ef657da))



## [1.0.1](http://github.com/forsigner/forsigner/egg-typeorm/compare/v1.0.0...v1.0.1) (2019-03-06)


### Bug Fixes

* check entity dir ([b4f891e](http://github.com/forsigner/forsigner/egg-typeorm/commit/b4f891e))


app.ts
import { join, sep } from 'path'
import { find } from 'fs-jetpack'
import { Application } from 'egg'
import {
  createConnection,
  getRepository,
  Connection,
  useContainer,
} from 'typeorm'
import { Container } from 'typedi'

import { watch } from 'chokidar'
import * as fs from 'fs-extra'
import * as prettier from 'prettier'

const hasTsLoader = typeof require.extensions['.ts'] === 'function';

export function formatCode(text: string) {
  return prettier.format(text, {
    semi: false,
    tabWidth: 2,
    singleQuote: true,
    parser: 'typescript',
    trailingComma: 'all',
  })
}

function handleConfig(config: any, _env: string) {
  if (hasTsLoader) {
    return config
  }
  const keys = ['entities', 'migrations', 'subscribers']
  for (const key of keys) {
    if (config[key]) {
      const newValue = config[key].map((item: string) =>
        item.replace(/\.ts$/, '.js'),
      )
      config[key] = newValue
    }
  }
  return config
}

async function connectDB(app: Application) {
  const config = handleConfig(app.config.typeorm, app.config.env)
  useContainer(Container)

  const connection = await createConnection(config)
  app.context.connection = connection
}

function capitalizeFirstLetter(str: string) {
  return str.charAt(0).toUpperCase() + str.slice(1)
}

function getModelName(file: string) {
  const filename = file.split(sep).pop() || ''
  const name = capitalizeFirstLetter(filename.replace(/\.ts$|\.js$/g, ''))
  return name
}

function writeTyping(path: string, text: string) {
  fs.writeFileSync(path, formatCode(text), { encoding: 'utf8' })
}

function getTypingText(
  importText: string,
  repoText: string,
  entityText: string,
) {
  const tpl = `
import 'egg'
import { Repository, Connection } from 'typeorm'
${importText}

declare module 'egg' {
  interface Context {
    connection: Connection
    entity: {
      ${entityText}
    }
    repo: {
      ${repoText}
    }
  }
}
`
  return tpl
}

function formatPaths(files: string[]) {
  return files.map(file => {
    const name = getModelName(file)
    file = file.split(sep).join('/')
    const importPath = `../${file}`.replace(/\.ts$|\.js$/g, '')
    return {
      name,
      importPath,
    }
  })
}

function watchEntity(app: Application) {
  const { baseDir } = app
  const entityDir = join(baseDir, 'app', 'entity')
  const typingsDir = join(baseDir, 'typings')

  if (!fs.existsSync(entityDir)) return

  fs.ensureDirSync(typingsDir)
  watch(entityDir).on('all', (eventType: string) => {
    if (['add', 'change'].includes(eventType)) {
      createTyingFile(app)
    }

    if (['unlink'].includes(eventType)) {
      createTyingFile(app)
    }
  })
}

function createTyingFile(app: Application) {
  const { baseDir } = app
  const entityDir = join(baseDir, 'app', 'entity')
  const files = find(entityDir, { matching: '*.ts' })
  const typingPath = join(baseDir, 'typings', 'typeorm.d.ts')
  const pathArr = formatPaths(files)
  const importText = pathArr
    .map(i => `import ${i.name} from '${i.importPath}'`)
    .join('\n')
  const repoText = pathArr
    .map(i => `${i.name}: Repository<${i.name}>`)
    .join('\n')

  // TODO
  const entityText = pathArr.map(i => `${i.name}: any`).join('\n')
  const text = getTypingText(importText, repoText, entityText)
  writeTyping(typingPath, text)
}

function getEntityFromPath(app: Application, entityPath: string) {
  const connection: Connection = app.context.connection
  const fileModule = require(entityPath)
  const entities = Object.keys(fileModule).reduce(
    (result, cur) => {
      try {
        // TODO: 太 hack
        connection.getMetadata(fileModule[cur])
        if (!result.includes(fileModule[cur])) {
          return [...result, fileModule[cur]]
        } else {
          return result
        }
      } catch {
        //
      }

      return result
    },
    [] as any[],
  )

  if (!entities.length) {
    throw new Error(`${entityPath} 格式不正确，不存在 @entity`)
  }

  return entities[0]
}

async function loadEntityAndModel(app: Application) {
  const { baseDir } = app
  const entityDir = join(baseDir, 'app', 'entity')

  if (!fs.existsSync(entityDir)) return

  const matching = hasTsLoader ? '*.ts' : '*.js'

  const files = find(entityDir, { matching })
  app.context.repo = {}
  app.context.entity = {}

  try {
    for (const file of files) {
      const entityPath = join(baseDir, file)
      const entity: any = getEntityFromPath(app, entityPath)
      const name = getModelName(file)
      app.context.repo[name] = getRepository(entity)
      app.context.entity[name] = entity
    }
  } catch (e) {
    app.logger.error(e)
  }
}

export default async (app: Application) => {
  const config = app.config.typeorm
  if (!config) {
    throw new Error('please config typeorm in config file')
  }

  app.beforeStart(async () => {
    try {
      await connectDB(app)
      // if (app.config.env === 'local') {
      watchEntity(app)
      // }
      await loadEntityAndModel(app)
      app.logger.info('[typeorm]', '数据链接成功')
    } catch (error) {
      app.logger.error('[typeorm]', '数据库链接失败')
      app.logger.error(error)
    }
  })
}


example/config/config.default.ts
import { EggAppConfig, EggAppInfo, PowerPartial } from 'egg'

export default (appInfo: EggAppInfo) => {
  const config = {} as PowerPartial<EggAppConfig>

  // override config from framework / plugin
  // use for cookie sign key, should change to your own and keep security
  config.keys = appInfo.name + '_1551875313892_8568'

  config.typeorm = {
    type: 'mysql',
    host: 'localhost',
    port: 3306,
    username: 'root',
    password: '123456',
    database: 'test',
    synchronize: true,
    logging: false,
    entities: ['app/entity/**/*.ts'],
    migrations: ['app/migration/**/*.ts'],
    subscribers: ['app/subscriber/**/*.ts'],
    cli: {
      entitiesDir: 'app/entity',
      migrationsDir: 'app/migration',
      subscribersDir: 'app/subscriber',
    },
  }

  // add your egg config in here
  config.middleware = []

  // add your special config in here
  const bizConfig = {
    sourceUrl: `https://github.com/eggjs/examples/tree/master/${appInfo.name}`,
  }

  // the return config will combines to EggAppConfig
  return {
    ...config,
    ...bizConfig,
  }
}
